Add infrastructure to download linux-firmwares.
Currently only libertas 88W8686 firmwares are handled.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2010-04-27 14:52:04.000000000 +0200
+++ buildroot/package/Config.in	2010-04-27 14:54:12.000000000 +0200
@@ -256,6 +256,7 @@
 source "package/e2fsprogs/Config.in"
 source "package/eeprog/Config.in"
 source "package/fconfig/Config.in"
+source "package/firmware/Config.in"
 source "package/fis/Config.in"
 source "package/fuse/Config.in"
 source "package/gadgetfs-test/Config.in"
Index: buildroot/package/firmware/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/firmware/Config.in	2010-04-28 14:57:55.000000000 +0200
@@ -0,0 +1,49 @@
+menuconfig BR2_PACKAGE_FIRMWARE
+	bool "linux-firmware"
+	help
+	  Install selected firmware from linux-firmware.git repository.
+
+	  This repository was setup by David Woodhouse in order to:
+	  "include firmware files from vendors who have chosen not to
+	  contribute their firmware to be distributed as part of the GPL'd
+	  kernel source tree, but would still like their firmware to be readily
+	  available to Linux users."
+
+
+if BR2_PACKAGE_FIRMWARE
+
+comment "Video"
+
+comment "WiFi chipsets"
+
+config BR2_PACKAGE_FIRMWARE_88W8686
+	bool "Marvell's 88W8686 (Libertas)"
+
+config BR2_PACKAGE_FIRMWARE_88W8686_SDIO
+	bool "SDIO"
+	default y
+	depends on BR2_PACKAGE_FIRMWARE_88W8686
+
+choice
+	prompt "Firmware version"
+	default BR2_PACKAGE_FIRMWARE_88W8686_SDIO_V9
+	depends on BR2_PACKAGE_FIRMWARE_88W8686_SDIO
+	help
+	  SDIO firmware is available in v8 or v9 version.
+
+config BR2_PACKAGE_FIRMWARE_88W8686_SDIO_V8
+	bool "v8"
+
+config BR2_PACKAGE_FIRMWARE_88W8686_SDIO_V9
+	bool "v9"
+
+endchoice
+
+config BR2_PACKAGE_FIRMWARE_88W8686_GSPI
+	bool "GSPI"
+	depends on BR2_PACKAGE_FIRMWARE_88W8686
+
+comment "Others"
+
+endif
+
Index: buildroot/package/firmware/firmware.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/firmware/firmware.mk	2010-04-28 17:28:13.000000000 +0200
@@ -0,0 +1,71 @@
+#############################################################
+#
+# Get and install some linux-firmware
+#
+#############################################################
+FIRMWARE_VERSION:=af5222c5ded5d944267acfbd001571409bea7eeb
+FIRMWARE_SOURCE:=firmware-$(FIRMWARE_VERSION).tar.gz
+FIRMWARE_GIT:=git://git.kernel.org/pub/scm/linux/kernel/git/dwmw2/linux-firmware.git
+FIRMWARE_DIR:=$(BUILD_DIR)/firmware-$(FIRMWARE_VERSION)
+FIRMWARE_TARGET_DIR:=$(TARGET_DIR)/lib/firmware
+FIRMWARE_CAT:=$(ZCAT)
+FIRMWARE_NAME:=firmware-$(FIRMWARE_VERSION)
+
+
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE_88W8686_SDIO)),y)
+FIRMWARE_BINARIES+=sd8686_helper.bin sd8686.bin
+endif
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE_88W8686_GSPI)),y)
+FIRMWARE_BINARIES+=gspi8686_helper.bin gspi8686.bin
+endif
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE_88W8686_SDIO_V8)),y)
+FIRMWARE_88W8686_SDIO_VERSION:=_v8
+else
+FIRMWARE_88W8686_SDIO_VERSION:=_v9
+endif
+
+$(DL_DIR)/$(FIRMWARE_SOURCE):
+	git clone $(FIRMWARE_GIT) $(FIRMWARE_DIR)
+	(cd $(FIRMWARE_DIR); \
+		git archive --prefix=$(FIRMWARE_NAME)/ $(FIRMWARE_VERSION) | \
+		gzip > $(DL_DIR)/$(FIRMWARE_SOURCE) )
+	rm -rf $(FIRMWARE_DIR)
+#	$(WGET) -O $(DL_DIR)/$(FIRMWARE_SOURCE) "$(FIRMWARE_URL)"
+
+$(FIRMWARE_DIR)/.unpacked: $(DL_DIR)/$(FIRMWARE_SOURCE)
+	$(FIRMWARE_CAT) $(DL_DIR)/$(FIRMWARE_SOURCE) | tar -C $(BUILD_DIR) $(TAR_OPTIONS) -
+	touch $@
+
+$(FIRMWARE_DIR)/.installed: $(FIRMWARE_DIR)/.unpacked
+	mkdir -p $(FIRMWARE_TARGET_DIR)
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE_88W8686_SDIO)),y)
+	$(INSTALL) -D $(FIRMWARE_DIR)/libertas/sd8686_v9_helper.bin $(FIRMWARE_TARGET_DIR)/sd8686_helper.bin
+	$(INSTALL) -D $(FIRMWARE_DIR)/libertas/sd8686_v9.bin $(FIRMWARE_TARGET_DIR)/sd8686.bin
+endif
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE_88W8686_GSPI)),y)
+	$(INSTALL) -D $(FIRMWARE_DIR)/libertas/gspi8686_v9_helper.bin $(FIRMWARE_TARGET_DIR)/gspi8686_helper.bin
+	$(INSTALL) -D $(FIRMWARE_DIR)/libertas/gspi8686_v9.bin $(FIRMWARE_TARGET_DIR)/gspi8686.bin
+endif
+	touch $@
+
+firmware: $(FIRMWARE_DIR)/.installed
+
+firmware-source: $(DL_DIR)/$(FIRMWARE_SOURCE)
+
+firmware-clean:
+	rm -f $(FIRMWARE_DIR)/.installed
+	(for file in $(FIRMWARE_BINARIES); do \
+		rm -f "$(FIRMWARE_TARGET_DIR)/$$file"; \
+	done)
+
+firmware-dirclean:
+	rm -rf $(FIRMWARE_DIR)
+
+#############################################################
+#
+# Toplevel Makefile options
+#
+#############################################################
+ifeq ($(strip $(BR2_PACKAGE_FIRMWARE)),y)
+TARGETS+=firmware
+endif
